modules/formulize/admin/formindex.php

```diff
@@ -609,13 +609,69 @@ function patch40() {
                 }
             }
         }
-        
+
+				// add opening <?php tags to code snippets that don't have them
+				// 1. derived value formulas $ele_value[0]
+				// 2. Textbox $ele_value[2] if they contain $default
+				// 3. Textarea $ele_value[0] if they contain $default
+				// 4. ib $ele_value[0] if they contain $value
+				// 5. areamodif $ele_value[0] if they contain $value
+				$elementsNeedingOpeningPHPTagsSQL = "SELECT ele_id, ele_type, ele_value FROM ".$xoopsDB->prefix('formulize')." WHERE
+					(ele_type = 'derived')
+					OR (ele_type IN ('ib', 'areamodif') AND ele_value LIKE '%\$value%')
+					OR (ele_type IN ('text', 'textarea') AND ele_value LIKE '%\$default%') ";
+				if($res = $xoopsDB->query($elementsNeedingOpeningPHPTagsSQL)) {
+					while($record = $xoopsDB->fetchArray($res)) {
+						$eleValueKey = 0;
+						if($record['ele_type'] == 'text') {
+							$eleValueKey = 2;
+						}
+						$newEleValue = unserialize(($record['ele_value']));
+						if(substr($newEleValue[$eleValueKey], 0, 5) != '<?php') {
+							$newEleValue[$eleValueKey] = "<?php\n".$newEleValue[$eleValueKey];
+							$newEleValue = serialize($newEleValue);
+							$updateSQL = "UPDATE ".$xoopsDB->prefix('formulize')." SET ele_value = ".$xoopsDB->quoteString($newEleValue)." WHERE ele_id = ".$record['ele_id'];
+							if(!$updateRes = $xoopsDB->query($updateSQL)) {
+								print "Notice: could not add opening PHP tag to the code in element ".$record['ele_id']." with the SQL:<br>".str_replace('<', '&lt;',$updateSQL)."<br>".$xoopsDB->error()."<br>This is not a critical error. You can add the tag yourself at the top of the code, if you want the editor to provide highlighting. For more information contact <a href=mailto:info@formulize.org>info@formulize.org</a>.<br>";
+							}
+						}
+					}
+				} else {
+					exit("Error detecting code snippets that need opening PHP tags. SQL dump:<br>".$elementsNeedingOpeningPHPTagsSQL."<br>".$xoopsDB->error()."<br>Please contact <a href=mailto:info@formulize.org>info@formulize.org</a> for assistance.");
+				}
+
+				// Same operation, on the form procedures
+				$formProceduresNeedingOpeningPHPTagsSQL = "SELECT id_form as fid, on_before_save, on_after_save, on_delete, custom_edit_check
```

Comment:
Split into separate operations/functions so it runs once for each event

Reply:
Wouldn't actually simplify things, because the multiple events are stored in the single record, so as written we retrieve the records that need to be updated, and then update the relevant events in the record and write it back to DB. Splitting up would potentially multiply the retrieving and writing by four. Not that we care much in the DB patch operation, but it's needless repetition. Added comments to capture the flow.